File: tests/overflow_check_test.py
#!/usr/bin/env -S uv run python3
"""overflow-check.py trigger/target-extraction regression tests. Run: uv run python3 tests/overflow_check_test.py"""
import importlib.util
import json
import subprocess
import sys
from pathlib import Path

HOOK_PATH = Path(__file__).resolve().parents[1] / "hooks" / "overflow-check.py"
spec = importlib.util.spec_from_file_location("overflow_check", HOOK_PATH)
overflow_check = importlib.util.module_from_spec(spec)
spec.loader.exec_module(overflow_check)  # noqa: E402

P, F = 0, 0


def check(name, cond, extra=""):
    global P, F
    if cond:
        P += 1; print(f"  ok  {name}")
    else:
        F += 1; print(f"  FAIL {name} {extra}")


resolve = overflow_check.resolve_typ_target

# (a) both compilers trigger; (b) flags before the target don't break extraction.
check("bare typst compile", resolve("typst compile slides.typ") == "slides.typ")
check("typst compile with --input first", resolve("typst compile --input handout=true slides.typ") == "slides.typ")
check("tinymist compile", resolve("tinymist compile slides.typ") == "slides.typ")
check("tinymist compile nested path", resolve("tinymist compile presentation/slides.typ") == "presentation/slides.typ")
check("typst compile --root form", resolve("typst compile --root . slides/lecture1.typ") == "slides/lecture1.typ")
check("cd prefix + typst compile", resolve("cd foo && typst compile slides.typ") == "slides.typ")
check("non-compile command: no trigger", resolve("typst watch slides.typ") is None)
check("unrelated command: no trigger", resolve("ls -la") is None)

# fail-open on malformed / non-Bash / non-triggering stdin.
def run_hook(stdin_text):
    return subprocess.run([sys.executable, str(HOOK_PATH)], input=stdin_text, capture_output=True, text=True, timeout=10)

check("malformed JSON -> exit 0", run_hook("not json").returncode == 0)
check("empty object -> exit 0", run_hook("{}").returncode == 0)
check("non-Bash tool -> exit 0", run_hook(json.dumps({"tool_name": "Read", "tool_input": {}})).returncode == 0)
check(
    "tinymist compile via Bash tool -> exit 0 (no crash even if plugin root / check script unresolved)",
    run_hook(json.dumps({"tool_name": "Bash", "tool_input": {"command": "tinymist compile slides.typ"}})).returncode == 0,
)

print(f"\n{P} passed, {F} failed")
sys.exit(1 if F else 0)

File: tests/overflow-check.test.mjs
// overflow-check trigger/target-extraction regression tests.
//
// PORTED FROM tests/overflow_check_test.py, WHICH HAD BEEN QUARANTINED. That file did
// `importlib.util.spec_from_file_location(... "hooks" / "overflow-check.py")` against a hook that
// is now TypeScript, so it could not even load — it was quarantined as "opens overflow-check.py;
// the hook is now .ts" and left there. Quarantining a suite because its SUBJECT moved retires the
// coverage silently: the hook kept shipping, and nothing checked it for as long as the entry stood.
//
// Run: bun tests/overflow-check.test.mjs
import { resolveTypTarget } from '../hooks/overflow-check.ts'

let PASS = 0, FAIL = 0
const ok = (name, condition, extra = '') => {
  if (condition) PASS++
  else { FAIL++; console.log(`FAIL  ${name}${extra ? ` — ${extra}` : ''}`) }
}

// (a) both compilers trigger; (b) flags BEFORE the target must not break extraction — the whole
// reason this suite exists is that a naive "last token" parse gets `--input handout=true` wrong.
ok('bare typst compile', resolveTypTarget('typst compile slides.typ') === 'slides.typ')
ok('typst compile with --input first', resolveTypTarget('typst compile --input handout=true slides.typ') === 'slides.typ')
ok('tinymist compile', resolveTypTarget('tinymist compile slides.typ') === 'slides.typ')
ok('tinymist compile nested path', resolveTypTarget('tinymist compile presentation/slides.typ') === 'presentation/slides.typ')
ok('typst compile --root form', resolveTypTarget('typst compile --root . slides/lecture1.typ') === 'slides/lecture1.typ')
ok('cd prefix + typst compile', resolveTypTarget('cd foo && typst compile slides.typ') === 'slides.typ')
ok('non-compile command: no trigger', resolveTypTarget('typst watch slides.typ') === null)
ok('unrelated command: no trigger', resolveTypTarget('ls -la') === null)

// Fail-open on malformed / non-Bash / non-triggering stdin. A PreToolUse hook that dies on junk
// input is a hook that blocks real work, so every one of these must exit 0.
const runHook = async stdin => {
  const proc = Bun.spawn(['bun', 'hooks/overflow-check.ts'], { stdin: 'pipe', stdout: 'pipe', stderr: 'pipe' })
  proc.stdin.write(stdin); proc.stdin.end()
  return await proc.exited
}

ok('malformed JSON -> exit 0', await runHook('not json') === 0)
ok('empty object -> exit 0', await runHook('{}') === 0)
ok('non-Bash tool -> exit 0', await runHook(JSON.stringify({ tool_name: 'Read', tool_input: {} })) === 0)
ok('tinymist compile via Bash tool -> exit 0 (no crash even if plugin root unresolved)',
   await runHook(JSON.stringify({ tool_name: 'Bash', tool_input: { command: 'tinymist compile slides.typ' } })) === 0)

console.log(`\n${PASS}/${PASS + FAIL} passed`)
if (FAIL) process.exit(1)
